File: src/core/multidimensional/array_descriptor.cpp
// SPDX-License-Identifier: GPL-3.0-only

#include <xmipp4/core/multidimensional/array_descriptor.hpp>

#include <boost/functional/hash.hpp>

#ifdef XMIPP4_CORE_ENABLE_RUST_BRIDGE
	#include "../rust_bridge.hpp"
#endif

namespace xmipp4 
{
namespace multidimensional
{

array_descriptor::array_descriptor() noexcept
	: m_data_type(numerical_type::unknown)
{
}

array_descriptor::array_descriptor(
	strided_layout layout, 
	numerical_type data_type
) noexcept
	: m_layout(std::move(layout))
	, m_data_type(data_type)
{
}

array_descriptor::array_descriptor(const array_descriptor &other) = default;
array_descriptor::array_descriptor(array_descriptor &&other) noexcept = default;
array_descriptor::~array_descriptor() = default;

array_descriptor&
array_descriptor::operator=(const array_descriptor &other) = default;
array_descriptor&
array_descriptor::operator=(array_descriptor &&other) noexcept = default;

std::size_t array_descriptor::hash() const noexcept
{
	auto seed = m_layout.hash();
	boost::hash_combine(seed, boost::hash_value(m_data_type));
	return seed;
}

const strided_layout& array_descriptor::get_layout() const noexcept
{
	return m_layout;
}

numerical_type array_descriptor::get_data_type() const noexcept
{
	return m_data_type;
}

bool operator==(
	const array_descriptor &lhs, 
	const array_descriptor &rhs
) noexcept
{
	return 
		lhs.get_data_type() == rhs.get_data_type() &&
		lhs.get_layout() == rhs.get_layout();
}

bool operator!=(
	const array_descriptor &lhs, 
	const array_descriptor &rhs
) noexcept
{
	return !(lhs == rhs);
}

bool is_initialized(const array_descriptor &descriptor) noexcept
{
	return descriptor.get_data_type() != numerical_type::unknown;
}

std::size_t compute_storage_requirement(const array_descriptor &descriptor)
{
	const auto &layout = descriptor.get_layout();
	const auto data_type = descriptor.get_data_type();

#ifdef XMIPP4_CORE_ENABLE_RUST_BRIDGE
	std::vector<std::size_t> extents;
	std::vector<std::ptrdiff_t> strides;
	layout.get_extents(extents);
	layout.get_strides(strides);

	const auto storage_requirement = xmipp4_rust_compute_storage_requirement(
		extents.data(),
		strides.data(),
		extents.size(),
		layout.get_offset()
	);
	return storage_requirement * get_size(data_type);
#else
	return layout.compute_storage_requirement() * get_size(data_type);
#endif
}

} // namespace multidimensional
} // namespace xmipp4

File: rust/core-types/tests/array_descriptor.rs
// SPDX-License-Identifier: GPL-3.0-only

use xmipp4_core_types::{
	compute_storage_requirement, is_initialized, ArrayDescriptor, NumericalType, StridedLayout,
};

#[test]
fn default_descriptor_is_not_initialized() {
	let descriptor = ArrayDescriptor::default();
	assert!(!is_initialized(&descriptor));
	assert_eq!(compute_storage_requirement(&descriptor), 0);
}

#[test]
fn initialized_depends_only_on_data_type() {
	let descriptor = ArrayDescriptor::new(StridedLayout::default(), NumericalType::Float32);
	assert!(is_initialized(&descriptor));
}

#[test]
fn descriptor_storage_requirement_uses_layout_and_data_type() {
	let layout = StridedLayout::make_contiguous_layout(vec![3, 4]);
	let descriptor = ArrayDescriptor::new(layout, NumericalType::Float32);
	assert!(is_initialized(&descriptor));
	assert_eq!(compute_storage_requirement(&descriptor), 3 * 4 * 4);
}

#[test]
fn unknown_data_type_has_zero_storage_requirement() {
	let layout = StridedLayout::make_contiguous_layout(vec![3, 4]);
	let descriptor = ArrayDescriptor::new(layout, NumericalType::Unknown);
	assert_eq!(compute_storage_requirement(&descriptor), 0);
}

#[test]
fn descriptor_equality_depends_on_type_and_layout() {
	let layout = StridedLayout::make_contiguous_layout(vec![2, 2]);
	let a = ArrayDescriptor::new(layout.clone(), NumericalType::Int16);
	let b = ArrayDescriptor::new(layout.clone(), NumericalType::Int16);
	let c = ArrayDescriptor::new(layout, NumericalType::UInt16);

	assert_eq!(a, b);
	assert_ne!(a, c);
}
